R/dplyr_methods.R

```diff
@@ -507,6 +507,208 @@ slice.SingleCellExperiment <- function(.data, ..., .by=NULL, .preserve=FALSE) {
     .data[, rownames(new_meta)]
 }
 
+#' @name slice_sample
+#' @rdname slice
+#' @inherit dplyr::slice_sample
+#' @examples
+#' data(pbmc_small)
+#' pbmc_small |> slice_sample(n=1)
+#' pbmc_small |> slice_sample(prop=0.1)
+#'
+#' @importFrom SummarizedExperiment colData
+#' @importFrom dplyr slice_sample
+#' @export
+slice_sample.SingleCellExperiment <- function(.data, ..., n=NULL,
+    prop=NULL, by=NULL, weight_by=NULL, replace=FALSE) {
+    lifecycle::signal_superseded("1.0.0", "sample_n()", "slice_sample()")
+
+    if (!is.null(n))
+        new_meta <-
+            .data |>
+            colData() |>
+            as_tibble(rownames=c_(.data)$name) |>
+            select(-everything(), c_(.data)$name, {{ by }}, {{ weight_by }}) |>
+            slice_sample(..., n=n, by={{ by }},
+                weight_by={{ weight_by }}, replace=replace)
+    else if (!is.null(prop))
+        new_meta <-
+            .data |>
+            colData() |>
+            as_tibble(rownames=c_(.data)$name) |>
+            select(-everything(), c_(.data)$name, {{ by }}, {{ weight_by }}) |>
+            slice_sample(..., prop=prop, by={{ by }},
+                weight_by={{ weight_by }}, replace=replace)
+    else
+        stop("tidySingleCellExperiment says:",
+            " you should provide `n` or `prop` arguments")
+
+    count_cells <- new_meta %>%
+        select(!!c_(.data)$symbol) %>%
+        count(!!c_(.data)$symbol)
+
+    # If repeated cells due to replacement
+    if (count_cells$n |> max() |> gt(1)){
+        message("tidySingleCellExperiment says: When sampling with replacement",
+            " a data frame is returned for independent data analysis.")
+        .data |>
+            as_tibble()  |>
+            right_join(new_meta %>% 
+                select(!!c_(.data)$symbol), by=c_(.data)$name)
+    } else {
+        .data[, pull(new_meta, !!c_(.data)$symbol)]
+    }
+}
+
+#' @name slice_head
+#' @rdname slice
+#' @inherit dplyr::slice_head
+#' @examples
+#' data(pbmc_small)
+#' # First rows based on existing order
+#' pbmc_small |> slice_head(n=5)
+#' 
+#' @importFrom dplyr slice_head
+#' @importFrom tibble rowid_to_column
+#' @export
+slice_head.SingleCellExperiment <- function(.data, ..., n, prop, by=NULL) {
+    row_number___ <- NULL
+    idx <- .data |>
+        colData() |>
+        as.data.frame() |>
+        select(-everything(), {{ by }}) |>
+        rowid_to_column(var='row_number___')  |>
+        slice_head(..., n=n, prop=prop, by={{ by }}) |>
+        pull(row_number___)
+
+    if (length(idx) == 0) {
+        stop("tidySingleCellExperiment says:",
+            " the resulting data container is empty.",
+            " Seurat does not allow for empty containers.")
+    }
```

Comment:
Look for "Seurat" in all files, and please replace ;)

Reply:
I believe an empty SCE is completely fine. Not sure this check is needed.

Reply:
Thank you very much for your prompt review. I apologize that the part of the message remained the same as `tidyseurat`. Also, to accept the empty SCE, all the checks for zero length ID in the slice functions are removed.